packages/reactive-element/src/decorators/custom-element.ts

```diff
@@ -13,6 +13,125 @@
+    getName: (ctor: CustomElementConstructor) => string | null;
+  }
+}
+
+export type DefineErrorCallback = (
+  error: Error,
+  tagName: string,
+  ctor: CustomElementConstructor
+) => void;
+
+/**
+ *
+ * @param tagName The tag name of the custom element to define.
+ * @param ctor The constructor of the custom element to define
+ * @param onError An optional function which can handle a define exception
+ * error. If provided, the exception is caught.
+ * @param onError
+ * @returns void
+ */
+export const define = (
+  tagName: string,
+  ctor: CustomElementConstructor,
+  onError?: DefineErrorCallback
+) => {
+  if (customElements.getName?.(ctor) === tagName) {
+    return;
+  }
+  try {
+    customElements.define(tagName, ctor);
+  } catch (error) {
+    if (onError === undefined) {
+      throw error;
+    } else {
+      onError?.(error as Error, tagName, ctor);
+    }
+  }
+};
+
+export const warn = (error: Error) => {
+  console.warn(error);
+};
+
+// https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+const semVerRe =
+  /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/;
+
+/**
+ * Parses a version string into an object providing semver version info.
+ * @param versionStr Version string to parse.
+ * @returns An object with keys for the semver info matching `vesrion`, `major`,
+ *  `minor`, `patch`, `prerelease`, and `buildmetadtata`.
+ */
+export const parseSemVer = (versionStr: string | undefined) => {
+  if (versionStr === undefined) {
+    return {};
+  }
+  const [version, major, minor, patch, prerelease, buildmetadata] =
+    versionStr.match(semVerRe) ?? [];
+  return {version, major, minor, patch, prerelease, buildmetadata};
+};
+
+export interface PackageVersion {
+  package?: string;
+  version?: string;
+}
+
+const UNKNOWN_PACKAGE = '<unknown package>';
+const UNKNOWN_VERSION = '<unknown version>';
+
+/**
+ * A define error callback which compares static `package` and `version`
+ * properties on the provided constructor against an already defined element
+ * of the given name and issues a console warning if the package or version
+ * is incompatible.
+ * @param error Exception error generated by the provoking define call.
+ * @param tagName The tag name provided in the provoking define call.
+ * @param ctor The constructor provided in the provoking define call.
+ * @returns
+ */
+export const warnPackageVersion = (
+  _error: Error,
+  tagName: string,
+  ctor: CustomElementConstructor
+) => {
+  const {package: definedPackage, version: definedVersion} =
+    (customElements.get(tagName) ?? {}) as PackageVersion;
+  const {package: newPackage, version: newVersion} = ctor as PackageVersion;
+  const packageOk = definedPackage === newPackage;
+  let versionOk = definedVersion === newVersion;
+  // versions match if using semver and "patch" is only diff.
```

Comment:
This is tricky. If the existing version is the same major and an equal or greater minor, they should be compatible.

But semver ranges are much more expressive, and without them we limit the expressiveness here. Pre-release versions do not follow the same rules either.

I'd be more comfortable if the semver stuff was left out of the first version, and maybe put in a labs package. We can vend a function that simply warns if the versions aren't exactly the same for now, or no default error handler at all.